{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0e009b4-5b08-4d9e-bbfe-db8f87cd77ca",
   "metadata": {},
   "source": [
    "# Technical assessment for Rancho Biosciences\n",
    "Emily Lorenzen "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "829bf25e-6d0c-4ddc-83ef-25c3ee525ea8",
   "metadata": {},
   "source": [
    "### 1. Set up project evironment\n",
    "- Create new conda environment \n",
    "- Installed pandas, openpyxl, and jupyter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "64fc7809-ea0c-48d0-823c-075577eaf752",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77c8d2f8-ecba-403c-a33d-8887590f6701",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Ask for file name, ask for sheet_name for each thing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b41ab1c-4bcc-4bfa-955a-dfe2d7bbf79b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Please create the report specified on the ‘Dat...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Worksheet</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Data Specification</td>\n",
       "      <td>Specification worksheet provides the column na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Patient  Clinical Data</td>\n",
       "      <td>Patient clinical data worksheet provides the p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tissue Sample Metadata</td>\n",
       "      <td>Tissue Sample Metadata is a sample QC report o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Serum Protein Data</td>\n",
       "      <td>Serum Protein data worksheet provides one assa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>RNA-seq (RPKM)</td>\n",
       "      <td>RNA-seq (RPKM) worksheet provides the second a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Example report</td>\n",
       "      <td>A few rows are provided to show the general fo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0  \\\n",
       "0  Please create the report specified on the ‘Dat...   \n",
       "1                                                NaN   \n",
       "2                                          Worksheet   \n",
       "3                                 Data Specification   \n",
       "4                             Patient  Clinical Data   \n",
       "5                             Tissue Sample Metadata   \n",
       "6                                 Serum Protein Data   \n",
       "7                                    RNA-seq (RPKM)    \n",
       "8                                     Example report   \n",
       "\n",
       "                                          Unnamed: 1  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                        Description  \n",
       "3  Specification worksheet provides the column na...  \n",
       "4  Patient clinical data worksheet provides the p...  \n",
       "5  Tissue Sample Metadata is a sample QC report o...  \n",
       "6  Serum Protein data worksheet provides one assa...  \n",
       "7  RNA-seq (RPKM) worksheet provides the second a...  \n",
       "8  A few rows are provided to show the general fo...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_excel('Technical Test - Data Wrangling.xlsx', engine = 'openpyxl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "547c8891-9cc9-46c1-8aae-4dae98641c1f",
   "metadata": {},
   "source": [
    "### 2. Import data from excel sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c896324-99c0-4714-bbc5-c58a49007b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_clinical_data = pd.read_excel('Technical Test - Data Wrangling.xlsx', sheet_name = 'Patient_clinical_data', engine = 'openpyxl')\n",
    "tissue_sample_metadata = pd.read_excel('Technical Test - Data Wrangling.xlsx', sheet_name = 'Tissue Sample Metadata')\n",
    "serum_protein_data = pd.read_excel('Technical Test - Data Wrangling.xlsx', sheet_name = 'Serum Protein data')\n",
    "rna_seq_data = pd.read_excel('Technical Test - Data Wrangling.xlsx', sheet_name = 'RNA-seq (RPKM)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9af2790e-9b9e-402f-9ef9-f42cf6515c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = [patient_clinical_data, tissue_sample_metadata, serum_protein_data, rna_seq_data]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f86650eb-e0ef-4886-b239-c37d759fc683",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Study_ID  Patient  Number Sex        Age\n",
      "0  RO4532A             1001   M  41.673240\n",
      "1  RO4532A             1002   M  53.200569\n",
      "2  RO4532A             1003   F  51.000000\n",
      "3  RO4532A             1004   F  66.229214\n",
      "4  RO4532A             1105   M  53.304328\n",
      "   Patient  Number Sample    Sample type Material  RIN  Total Reads(millions)\n",
      "0             1001  CRC1N         Normal      RNA    7                 50.444\n",
      "1             1001  CRC1T    Liver Tumor      RNA    9                 38.997\n",
      "2             1001  CRC1M  Metastic Lung      RNA    9                 45.367\n",
      "3             1002  CRC2N         Normal      RNA    7                 38.673\n",
      "4             1002  CRC2T    Liver Tumor      RNA    9                 41.477\n",
      "   Patient    Sample Serum IL-6 (g/L) Serum IL-6 Receptor (mg/L)\n",
      "0     1001  S-CRC1-A             73.7                       0.97\n",
      "1     1001  S-CRC1-B             67.7                       1.49\n",
      "2     1001  S-CRC1-C             71.3                        1.8\n",
      "3     1002  S-CRC2-A              QNS                        QNS\n",
      "4     1002  S-CRC2-B             33.1                      <LLOQ\n",
      "  GeneID      CRC1N     CRC1T      CRC1M      CRC2N      CRC2T      CRC2M  \\\n",
      "0  ICAM1   4.662774  8.354152  11.863796   5.501785  10.014376  12.980251   \n",
      "1    IL6   0.000000  0.783566   0.874109   0.000000   2.517168   0.683496   \n",
      "2   IL6R  12.153740  3.118068   5.728374  12.732372   1.808001   3.215843   \n",
      "3  VCAM1   4.596351  3.168585   9.143465   4.827985   6.111611   3.900552   \n",
      "4   SELE   0.077086  1.496338   1.779753   0.099606   2.355495   0.210465   \n",
      "\n",
      "       CRC3N      CRC3T      CRC3M  ...      CRC4T     CRC4M     CRC5T  \\\n",
      "0   6.334321  10.880375  13.788794  ...  15.089534  9.005161  6.829498   \n",
      "1   0.108353   0.443711   0.000000  ...   0.358051  0.215563  0.132953   \n",
      "2  10.183599   4.035287   1.545560  ...   3.312816  4.748433  2.888519   \n",
      "3  10.959019   7.771770   3.187510  ...   8.829485  2.862093  7.185174   \n",
      "4   0.042897   1.014963   0.051518  ...   0.859222  0.778481  0.326909   \n",
      "\n",
      "      CRC5N      CRC5M     CRC8N     CRC8T    CRC10M    CRC10N     CRC10T  \n",
      "0  4.872094  10.344830  8.496029  6.433474  2.482575  9.459477  10.449776  \n",
      "1  0.253945   0.618038  0.280569  0.820672  1.797076  0.000000   0.068515  \n",
      "2  1.478570   3.644413  3.639156  3.862186  3.701024  3.544909   0.693627  \n",
      "3  6.582593   4.299935  7.068253  5.343884  2.688047  7.137615   7.674014  \n",
      "4  0.749328   0.211455  0.767408  0.468175  0.827498  0.996252   0.712139  \n",
      "\n",
      "[5 rows x 21 columns]\n"
     ]
    }
   ],
   "source": [
    "for df in df_list:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1759ff2-f279-482c-871d-c0f853fe4534",
   "metadata": {},
   "source": [
    "### 3. Add additional columns to dataframes and change datatypes as needed\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6c608a-eb5c-4c95-8090-69a88f16a809",
   "metadata": {},
   "source": [
    "#### 3-1. Changes made to the patient_clinical_data dataframe\n",
    "- Add Unique_Patient_ID column to patient_clinical data\n",
    "- Round patient age to nearest number and convert to integer\n",
    "- Convert sex to Male and Female instead of M and F\n",
    "- Rename Patient Number column to Patient_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "624122d9-282c-49b5-8ee8-defdf53ca912",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Unique_Patient_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1002</td>\n",
       "      <td>MALE</td>\n",
       "      <td>53</td>\n",
       "      <td>RO4532A_1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1003</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>51</td>\n",
       "      <td>RO4532A_1003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1004</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>66</td>\n",
       "      <td>RO4532A_1004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1105</td>\n",
       "      <td>MALE</td>\n",
       "      <td>53</td>\n",
       "      <td>RO4532A_1105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Study_ID  Patient_ID     Sex  Age Unique_Patient_ID\n",
       "0  RO4532A        1001    MALE   42      RO4532A_1001\n",
       "1  RO4532A        1002    MALE   53      RO4532A_1002\n",
       "2  RO4532A        1003  FEMALE   51      RO4532A_1003\n",
       "3  RO4532A        1004  FEMALE   66      RO4532A_1004\n",
       "4  RO4532A        1105    MALE   53      RO4532A_1105"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add unique_patient_id column to patient_clinical_data df\n",
    "patient_clinical_data['Unique_Patient_ID'] = patient_clinical_data['Study_ID'] + '_' + patient_clinical_data['Patient  Number'].astype(str)\n",
    "\n",
    "# Round age to nearest number and convert to integer datatype\n",
    "patient_clinical_data.Age = patient_clinical_data.Age.round(decimals = 0).astype(int)\n",
    "\n",
    "# Convert sex to say male or female instead of M or F \n",
    "patient_clinical_data.Sex = patient_clinical_data.Sex.replace({'M': 'MALE', 'F': 'FEMALE'})\n",
    "\n",
    "#Rename Patient Number column to Patient_ID and convert to integer\n",
    "patient_clinical_data.rename({'Patient  Number': 'Patient_ID'}, axis = 1, inplace = True)\n",
    "patient_clinical_data.Patient_ID.astype(int)\n",
    "\n",
    "patient_clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "273874bc-5a53-48af-be82-3b0e997e17d6",
   "metadata": {},
   "source": [
    "#### 3-2 Changes made to the tissue_sample_metadata datafram\n",
    "- Add results_units\n",
    "- Drop RIN and Total Read(millions) columns\n",
    "- Rename columns: Sample type to Sample_General_Pathology, Sample to Sample_ID, Material to Material_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "952ece2d-7ceb-432f-b33d-7a8e9abcb220",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_General_Pathology</th>\n",
       "      <th>Material_type</th>\n",
       "      <th>Result_Units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>Normal</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RPKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1T</td>\n",
       "      <td>Liver Tumor</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RPKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>CRC1M</td>\n",
       "      <td>Metastic Lung</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RPKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002</td>\n",
       "      <td>CRC2N</td>\n",
       "      <td>Normal</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RPKM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002</td>\n",
       "      <td>CRC2T</td>\n",
       "      <td>Liver Tumor</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RPKM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient_ID Sample_ID Sample_General_Pathology Material_type Result_Units\n",
       "0        1001     CRC1N                   Normal           RNA         RPKM\n",
       "1        1001     CRC1T              Liver Tumor           RNA         RPKM\n",
       "2        1001     CRC1M            Metastic Lung           RNA         RPKM\n",
       "3        1002     CRC2N                   Normal           RNA         RPKM\n",
       "4        1002     CRC2T              Liver Tumor           RNA         RPKM"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add results_units column to tissue_sample_metadata\n",
    "tissue_sample_metadata['Result_Units'] = 'RPKM'\n",
    "\n",
    "# Drop RUN and Total Read(millions) columnb\n",
    "tissue_sample_metadata.drop(labels = ['RIN', 'Total Reads(millions)'], axis = 1, inplace = True)\n",
    "\n",
    "#Rename Sample type column\n",
    "tissue_sample_metadata.rename({'Sample type': 'Sample_General_Pathology', 'Sample': 'Sample_ID', 'Material': 'Material_type', 'Patient  Number': 'Patient_ID'}, axis = 1, inplace = True)\n",
    "\n",
    "tissue_sample_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b513774-7351-46f1-8564-0a35bed1f64d",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3-3 Changes made to the rna_seq_data dataframe\n",
    "- Transpose dataframe\n",
    "- Isolate the ICAM1 column as a series\n",
    "- Add Gene_Symbol column \n",
    "- Rename ICAM1 column as Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "50131168-4ca7-45be-8e7c-d80cec5dd6b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CRC1N      4.662774\n",
       "CRC1T      8.354152\n",
       "CRC1M     11.863796\n",
       "CRC2N      5.501785\n",
       "CRC2T     10.014376\n",
       "CRC2M     12.980251\n",
       "CRC3N      6.334321\n",
       "CRC3T     10.880375\n",
       "CRC3M     13.788794\n",
       "CRC4N      7.187844\n",
       "CRC4T     15.089534\n",
       "CRC4M      9.005161\n",
       "CRC5T      6.829498\n",
       "CRC5N      4.872094\n",
       "CRC5M      10.34483\n",
       "CRC8N      8.496029\n",
       "CRC8T      6.433474\n",
       "CRC10M     2.482575\n",
       "CRC10N     9.459477\n",
       "CRC10T    10.449776\n",
       "Name: ICAM1, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq_data_t = rna_seq_data.transpose()\n",
    "rna_seq_data_t = rna_seq_data_t.rename(columns=rna_seq_data_t.iloc[0]).drop(rna_seq_data_t.index[0])\n",
    "rna_seq_icam1 = rna_seq_data_t.loc[:, 'ICAM1']\n",
    "rna_seq_icam1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60ae0865-0d60-4fb9-99e6-9efe701b3097",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transpose rna_seq_data \n",
    "rna_seq_data_t = rna_seq_data.transpose()\n",
    "\n",
    "# Name columns based on first row information (currently the name of the genes)\n",
    "rna_seq_data_t = rna_seq_data_t.rename(columns=rna_seq_data_t.iloc[0]).drop(rna_seq_data_t.index[0])\n",
    "\n",
    "gene_list = rna_seq_data_t.columns\n",
    "#Isolate ICAM1 column\n",
    "\n",
    "df_list = []\n",
    "for gene in gene_list:\n",
    "    rna_seq_gene = rna_seq_data_t.loc[:, [gene]]\n",
    "\n",
    "    #Create Gene_Symbol column\n",
    "    rna_seq_gene['Gene_Symbol'] = gene\n",
    "\n",
    "    #Rename ICAM1 column to Results\n",
    "    rna_seq_gene.rename({gene: 'Result'}, axis = 1, inplace = True)\n",
    "    \n",
    "    df_list.append(rna_seq_gene)\n",
    "rna_seq_all = pd.concat(df_list)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd72bd8-6eac-4f15-801e-34df4f69a291",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#### 3-4 Changes made to the serum_protein_data dataframe\n",
    "- Add material type column\n",
    "- Convert data in the Serum IL-6 Receptor and Serum IL-6 columns to floats coercing non-float numbers to NaN\n",
    "- Recalculate IL-6 receptor levels as g/L instead of mg/L \n",
    "- Drop the IL-6 receptor mg/L column\n",
    "- Add in Result_Units column\n",
    "- Rename Serum IL-6 and Serum IL-6 Receptor columns to IL6 and IL6R, respectively. Rename Patient to Patient_ID and Sample to Sample_ID\n",
    "- ******Increase number of digits shown in the IL6 column\n",
    "- Use melt to convert the Serum IL-6 and Serum IL-6 Receptor columns into rows under the Gene_Symbol column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "57bf8716-87f1-48d4-a1be-1bbe25accc4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Result_Units</th>\n",
       "      <th>Material_type</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-A</td>\n",
       "      <td>g/L</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>73.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-B</td>\n",
       "      <td>g/L</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>67.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>S-CRC1-C</td>\n",
       "      <td>g/L</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>71.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1002</td>\n",
       "      <td>S-CRC2-A</td>\n",
       "      <td>g/L</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002</td>\n",
       "      <td>S-CRC2-B</td>\n",
       "      <td>g/L</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>33.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient_ID Sample_ID Result_Units Material_type Gene_Symbol  Result\n",
       "0        1001  S-CRC1-A          g/L         Serum         IL6    73.7\n",
       "1        1001  S-CRC1-B          g/L         Serum         IL6    67.7\n",
       "2        1001  S-CRC1-C          g/L         Serum         IL6    71.3\n",
       "3        1002  S-CRC2-A          g/L         Serum         IL6     NaN\n",
       "4        1002  S-CRC2-B          g/L         Serum         IL6    33.1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Add Material_type column\n",
    "serum_protein_data['Material_type'] = 'Serum'\n",
    "\n",
    "#Convert data type of the Serum IL-6 Receptor and Serum IL-6 column from string to float, using error coercion\n",
    "serum_protein_data['Serum IL-6 Receptor (g/L)'] = pd.to_numeric(serum_protein_data['Serum IL-6 Receptor (mg/L)'], errors = 'coerce')\n",
    "\n",
    "serum_protein_data['Serum IL-6 (g/L)'] = pd.to_numeric(serum_protein_data['Serum IL-6 (g/L)'], errors = 'coerce')\n",
    "\n",
    "#Convert mg/L unit to g/L for the Serum IL-6 receptor column in serum_protein_data dataframe\n",
    "serum_protein_data['Serum IL-6 Receptor (g/L)'] = serum_protein_data['Serum IL-6 Receptor (g/L)'].apply(lambda x: x * 0.001)\n",
    "\n",
    "#Remove Serum IL-6 Receptor Column with mg/L units, since it has now been calculated as g/L units\n",
    "serum_protein_data.drop(labels = 'Serum IL-6 Receptor (mg/L)', axis = 1, inplace = True)\n",
    "\n",
    "#Add Result_Units column\n",
    "serum_protein_data['Result_Units'] = 'g/L'\n",
    "\n",
    "#Rename Serum IL-6 and Serum IL-6 Receptor columns to IL6 and IL6R\n",
    "serum_protein_data.rename({'Serum IL-6 Receptor (g/L)': 'IL6R', 'Serum IL-6 (g/L)': 'IL6', 'Patient': 'Patient_ID', 'Sample': 'Sample_ID'}, axis = 1, inplace = True)\n",
    "\n",
    "# Melt the IL6 and IL6R columns into rows\n",
    "serum_protein_data = serum_protein_data.melt(id_vars = ['Patient_ID', 'Sample_ID', 'Result_Units', 'Material_type'], var_name = 'Gene_Symbol', value_name = 'Result')\n",
    "\n",
    "serum_protein_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13524cdc-aa1b-4358-ae6b-38b6d2bab5ac",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 4. Merge and concatenate the dataframes. \n",
    "- Merge patient_clinical_data with tissue_sample_metadata to make a clinical_tissue dataframe\n",
    "- Merge tissue clinical_tissue dataframe with rna_seq_icam1 dataframe to make clinical_tissue_rnaseq dataframe\n",
    "- Merge patient_clinical_data with serum_protein_data to make clinical_serum dataframe\n",
    "- Concatenate clinical_tissue_rnaseq dataframe with serum_protein_data dataframe to make clinical_tissue_rnaseq_serum \n",
    "dataframe\n",
    "- Add status column based on if result column is null or not\n",
    "- Organize dataframe so that patient information is grouped together "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "af26722e-6f5c-4c16-9fd3-aecc926778ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge tissue_sample df with patient_clinical_data df\n",
    "clinical_tissue = patient_clinical_data.merge(tissue_sample_metadata)\n",
    "\n",
    "# Merge rna_seq df with clinical_tissue df\n",
    "clinical_tissue_rnaseq = clinical_tissue.merge(rna_seq_icam1, left_on = 'Sample_ID', right_index = True)\n",
    "\n",
    "# Merge patient_clinical_data df with serum_protein_data df b\n",
    "clinical_serum = patient_clinical_data.merge(serum_protein_data)\n",
    "\n",
    "# Concatenate clinical_serum and clinical_tissue_rnaseq dataframes\n",
    "all_clinical_info = pd.concat([clinical_serum, clinical_tissue_rnaseq])\n",
    "\n",
    "# Create new column for status based on if results are NaN or numeric\n",
    "all_clinical_info['Status'] = np.where(all_clinical_info.Result.isnull(), 'Not Done', np.NaN)\n",
    "\n",
    "#Finally sort the dataframe\n",
    "all_clinical_info.sort_values(['Patient_ID', 'Material_type', 'Sample_ID', 'Gene_Symbol'], inplace = True)\n",
    "\n",
    "# Get columns into the correct order \n",
    "col_list = all_clinical_info.columns.tolist()\n",
    "col_order = ['Study_ID', 'Patient_ID','Sex','Age','Unique_Patient_ID','Sample_ID','Sample_General_Pathology','Material_type','Gene_Symbol','Result', 'Result_Units', 'Status']\n",
    "all_clinical_info.reindex(columns = col_order, inplace = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c5565c64-336b-431e-8e75-6975132e0c05",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Write final dataframe into original excel file as a new sheet\n",
    "\n",
    "# Identify file to create a new sheet in\n",
    "all_clinical_info.to_excel('Technical Test Submission - Data Wrangling.xlsx', 'Final Database', na_rep = 'NaN', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "55ad0f5c-dfc8-4c32-ac07-2103773d1dff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Study_ID',\n",
       " 'Patient_ID',\n",
       " 'Sex',\n",
       " 'Age',\n",
       " 'Unique_Patient_ID',\n",
       " 'Sample_ID',\n",
       " 'Result_Units',\n",
       " 'Material_type',\n",
       " 'Gene_Symbol',\n",
       " 'Result',\n",
       " 'Sample_General_Pathology',\n",
       " 'ICAM1',\n",
       " 'Status']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "col_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "86876102-7fcc-4a81-928a-fad0a8e989e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_order = ['Study_ID', 'Patient_ID','Sex','Age','Unique_Patient_ID','Sample_ID','Sample_General_Pathology','Material_type','Gene_Symbol','Result', 'Result_Units', 'Status']\n",
    "\n",
    " \n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28f04bed-9341-4187-831d-94c6060b4070",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>Unique_Patient_ID</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Sample_General_Pathology</th>\n",
       "      <th>Material_type</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Result</th>\n",
       "      <th>Result_Units</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>CRC1M</td>\n",
       "      <td>Metastic Lung</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>Not Done</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>CRC1N</td>\n",
       "      <td>Normal</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>Not Done</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>CRC1T</td>\n",
       "      <td>Liver Tumor</td>\n",
       "      <td>RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPKM</td>\n",
       "      <td>Not Done</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>S-CRC1-A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>73.70000</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>1001</td>\n",
       "      <td>MALE</td>\n",
       "      <td>42</td>\n",
       "      <td>RO4532A_1001</td>\n",
       "      <td>S-CRC1-A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>0.00097</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2311</td>\n",
       "      <td>MALE</td>\n",
       "      <td>75</td>\n",
       "      <td>RO4532A_2311</td>\n",
       "      <td>S-CRC11-a</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>0.00345</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2311</td>\n",
       "      <td>MALE</td>\n",
       "      <td>75</td>\n",
       "      <td>RO4532A_2311</td>\n",
       "      <td>S-CRC11-b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>37.60000</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2311</td>\n",
       "      <td>MALE</td>\n",
       "      <td>75</td>\n",
       "      <td>RO4532A_2311</td>\n",
       "      <td>S-CRC11-b</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>0.00301</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2311</td>\n",
       "      <td>MALE</td>\n",
       "      <td>75</td>\n",
       "      <td>RO4532A_2311</td>\n",
       "      <td>S-CRC11-c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6</td>\n",
       "      <td>31.40000</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>RO4532A</td>\n",
       "      <td>2311</td>\n",
       "      <td>MALE</td>\n",
       "      <td>75</td>\n",
       "      <td>RO4532A_2311</td>\n",
       "      <td>S-CRC11-c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Serum</td>\n",
       "      <td>IL6R</td>\n",
       "      <td>0.00294</td>\n",
       "      <td>g/L</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Study_ID  Patient_ID   Sex  Age Unique_Patient_ID  Sample_ID  \\\n",
       "2   RO4532A        1001  MALE   42      RO4532A_1001      CRC1M   \n",
       "0   RO4532A        1001  MALE   42      RO4532A_1001      CRC1N   \n",
       "1   RO4532A        1001  MALE   42      RO4532A_1001      CRC1T   \n",
       "0   RO4532A        1001  MALE   42      RO4532A_1001   S-CRC1-A   \n",
       "3   RO4532A        1001  MALE   42      RO4532A_1001   S-CRC1-A   \n",
       "..      ...         ...   ...  ...               ...        ...   \n",
       "45  RO4532A        2311  MALE   75      RO4532A_2311  S-CRC11-a   \n",
       "43  RO4532A        2311  MALE   75      RO4532A_2311  S-CRC11-b   \n",
       "46  RO4532A        2311  MALE   75      RO4532A_2311  S-CRC11-b   \n",
       "44  RO4532A        2311  MALE   75      RO4532A_2311  S-CRC11-c   \n",
       "47  RO4532A        2311  MALE   75      RO4532A_2311  S-CRC11-c   \n",
       "\n",
       "   Sample_General_Pathology Material_type Gene_Symbol    Result Result_Units  \\\n",
       "2             Metastic Lung           RNA         NaN       NaN         RPKM   \n",
       "0                    Normal           RNA         NaN       NaN         RPKM   \n",
       "1               Liver Tumor           RNA         NaN       NaN         RPKM   \n",
       "0                       NaN         Serum         IL6  73.70000          g/L   \n",
       "3                       NaN         Serum        IL6R   0.00097          g/L   \n",
       "..                      ...           ...         ...       ...          ...   \n",
       "45                      NaN         Serum        IL6R   0.00345          g/L   \n",
       "43                      NaN         Serum         IL6  37.60000          g/L   \n",
       "46                      NaN         Serum        IL6R   0.00301          g/L   \n",
       "44                      NaN         Serum         IL6  31.40000          g/L   \n",
       "47                      NaN         Serum        IL6R   0.00294          g/L   \n",
       "\n",
       "      Status  \n",
       "2   Not Done  \n",
       "0   Not Done  \n",
       "1   Not Done  \n",
       "0        nan  \n",
       "3        nan  \n",
       "..       ...  \n",
       "45       nan  \n",
       "43       nan  \n",
       "46       nan  \n",
       "44       nan  \n",
       "47       nan  \n",
       "\n",
       "[68 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_clinical_info.reindex(columns = col_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b79438-ee92-4d1b-8236-5afb0d789dc8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
